### DISCUSSION TITLE
[Feature Request] MacOS support to collect Antivirus information

### DISCUSSION THREAD

**Discussion by tiangao88** [glpi-project/glpi-agent#367] [Ideas]
### Your idea

Are there any plan to support collecting antivirus information on MacOS?
Is it technically possible?

So far the antivirus information is collected as software and available in GLPI but we found no way to requalify it as antivirus.
Thanks!

**Comment by g-bougard** [comment]
Hi @tiangao88 

there's actually no plan to add such support. And indeed, you're the first person I see interested with that feature on MacOSX ;-)

This is surely technically possible. To start such an feature, we probably need a list of the major AV provider on MacOSX and we need to know if they at last provide a free or evaluation version so we can try them in a dedicated VM. 

Do you know if `system_profiler` command gives any installed AV information ?

**Reply by tiangao88** [reply]
Hi,

Sorry for being such a noob but I couldn't find in the man pages how to run the _system_profiler_ command with _glpi-inventory_ or _glpi-agent_. Can you give me instructions?

_glpi-inventory --partial antivirus_ returns no information about antivirus

_glpi-inventory --partial software_ returns this regarding the antivirus we use
         _{
            "install_date": "2023-02-21",
            "name": "Microsoft Defender",
            "system_category": "Applications",
            "version": "101.97.94"
         }_


Here is the information about Microsoft Defender for Endpoint on Mac
https://learn.microsoft.com/en-us/microsoft-365/security/defender-endpoint/microsoft-defender-endpoint-mac?view=o365-worldwide

I am afraid it requires a subscription for use but probably does not block the installation.

On more noob question: it seems we don't need plugins on GLPI server side to receive the inventory from GPLI Agent so what is the plugin for?
https://plugins.glpi-project.org/#/plugin/glpiinventory

**Comment by tiangao88** [comment]
Hi @g-bougard 

So I figured `system_profile` was an OS command and indeed I see traces of Microsoft Defender in the output

```
      Microsoft Defender:
      Version: 101.98.30
      Obtained from: Identified Developer
      Last Modified: 30/03/2023 18:12
      Kind: Universal
      Signed by: Developer ID Application: Microsoft Corporation (UBF8T346G9), Developer ID Certification Authority, Apple Root CA
      Location: /Applications/Microsoft Defender.app
```

There are also traces of all updates

```
      Microsoft Defender:
      Version: 101.96.85
      Source: 3rd Party
      Install Date: 23/01/2023 20:02

    Microsoft Defender:
      Version: 101.97.94
      Source: 3rd Party
      Install Date: 21/02/2023 14:56

    Microsoft Defender:
      Version: 101.98.30
      Source: 3rd Party
      Install Date: 30/03/2023 18:12
```

There is no specific section on Antivirus in the output though.

I have also installed GLPI Inventory Plugin to try to collect file information from the Mac but it does not work so well.

Is there any chance you put detecting antivirus on Mac on the roadmap for GLPI Agent?

Thanks.

**Comment by g-bougard** [comment]
Hi @tiangao88 
do you have this information in the installed software section ?
Here this only show the AV is installed but we don't know if it's enabled. Also it would be nice to know if it's using the latest virus signature database.
Just checking the online I can see we may have a simple way to catch the required information.
So from a root terminal can you check if you have an output for this command: `mdatp health`
If yes, can you share it ?

**Comment by tiangao88** [comment]
Hi @g-bougard ,

Indeed the information about Microsoft Defender is in the `Applications:` section.
Here is the output of `mdatp health` 

```
healthy                                     : true
health_issues                               : []
licensed                                    : true
engine_version                              : "1.1.20100.6"
app_version                                 : "101.98.30"
org_id                                      : "***anonymized***"
log_level                                   : "info"
machine_guid                                : "***anonymized***"
release_ring                                : "Production"
product_expiration                          : Sep 06, 2023 at 07:12:28 AM
cloud_enabled                               : true
cloud_automatic_sample_submission_consent   : "safe"
cloud_diagnostic_enabled                    : true
passive_mode_enabled                        : false
real_time_protection_enabled                : true
real_time_protection_available              : true
real_time_protection_subsystem              : "endpoint_security_extension"
network_events_subsystem                    : "network_filter_extension"
device_control_enforcement_level            : "audit"
tamper_protection                           : "block"
automatic_definition_update_enabled         : true
definitions_updated                         : May 02, 2023 at 06:26:13 PM
definitions_updated_minutes_ago             : 127
definitions_version                         : "1.387.2964.0"
definitions_status                          : "up_to_date"
edr_early_preview_enabled                   : "disabled"
edr_device_tags                             : []
edr_group_ids                               : ""
edr_configuration_version                   : "20.199999.main.2023.04.25.09-***anonymized***"
edr_machine_id                              : "***anonymized***"
conflicting_applications                    : []
network_protection_status                   : "stopped"
network_protection_enforcement_level        : "disabled"
data_loss_prevention_status                 : "disabled"
full_disk_access_enabled                    : true
```

**Reply by g-bougard** [reply]
Thank you, that's finally a nice start. I started to work on a dedicated module.
I noticed it should be possible to obtain a JSON output which could be far better to analyze.
So can you try and share the output of following command ?
```
mdatp health -output json
```

**Reply by tiangao88** [reply]
Hi @g-bougard ,

Here is the exact output of `mdatp health --output json`

```
{"$type":"getHealthAllResult","appVersion":"101.98.30","automaticDefinitionUpdateEnabled":{"displayValue":true,"isManaged":false,"value":true},"cloudAutomaticSampleSubmissionConsent":{"displayValue":true,"isManaged":false,"value":"safe"},"cloudDiagnosticEnabled":{"displayValue":true,"isManaged":false,"value":true},"cloudEnabled":{"displayValue":true,"isManaged":false,"value":true},"conflictingApplications":[],"dataLossPreventionStatus":"disabled","definitionsStatus":{"$type":"upToDate"},"definitionsUpdated":"1683092938089","definitionsUpdatedMinutesAgo":158,"definitionsVersion":"1.389.10.0","deviceControlEnforcementLevel":"audit","edrConfigurationVersion":"20.199999.main.2023.04.25.09-***anonymized***","edrDeviceTags":{"displayValue":true,"isManaged":false,"value":[]},"edrEarlyPreviewEnabled":{"displayValue":true,"isManaged":false,"value":"disabled"},"edrGroupIds":{"displayValue":true,"isManaged":false,"value":""},"edrMachineId":"***anonymized***","engineVersion":"1.1.20100.6","fullDiskAccessEnabled":true,"functionalityMode":{"displayValue":false,"isManaged":false,"value":"default"},"healthIssues":[],"healthy":true,"licensed":true,"logLevel":"info","machineGuid":"***anonymized***","networkEventsSubsystem":"network_filter_extension","networkProtectionEnforcementLevel":{"displayValue":true,"isManaged":false,"value":"disabled"},"networkProtectionStatus":"stopped","orgId":"***anonymized***","passiveModeEnabled":{"displayValue":true,"isManaged":false,"value":false},"productExpiration":"1693977148000","realTimeProtectionAvailable":true,"realTimeProtectionEnabled":{"displayValue":true,"isManaged":false,"value":true},"realTimeProtectionSubsystem":"endpoint_security_extension","releaseRing":"Production","tamperProtection":{"displayValue":true,"isManaged":false,"value":"block"},"telemetryClient":{"displayValue":false,"isManaged":false,"value":"OneDs"}}
```

**Reply by g-bougard** [reply]
Thank you, that's better for the expiration time provided: it's a timestamp and so safer to analyze.

**Reply by g-bougard** [reply]
Then can you copy as admin the file contained in this attached [AntiVirus.pm.zip](https://github.com/glpi-project/glpi-agent/files/11381031/AntiVirus.pm.zip) archive to the following location ? 
```
/Applications/GLPI-Agent/agent/GLPI/Agent/Task/Inventory/MacOS/AntiVirus.pm
```
Then run an inventory and tell us if you obtain the expected result in GLPI.

If that doesn't work as expected, you can share the output of:
```
sudo /Applications/GLPI-Agent/bin/glpi-inventory --partial antivirus --debug --debug
```

**Reply by tiangao88** [reply]
Hi @g-bougard 

Unfortunately the Antivirus section of the Computer Form is still empty.

<img width="867" alt="image" src="https://user-images.githubusercontent.com/53409436/235919760-1cff52f0-219c-41a1-b330-0b2537b37b98.png">


The inventory is made locally but the antivirus information seems to not be sent to GLPI server or maybe not processed by the server.

In doubt I installed the latest nightly build of the agent but it still does not work.

Attached the output of `sudo /Applications/GLPI-Agent/bin/glpi-inventory --partial antivirus --debug --debug`
[glpi-antivirus-inventory.txt](https://github.com/glpi-project/glpi-agent/files/11382591/glpi-antivirus-inventory.txt)

And here is the configuration of GLPI Agent

```
server = https://glpi.imyserver.com/front/inventory.php
ssl-fingerprint = **anonymized**
no-ssl-check = 0
timeout = 30
delaytime = 3600
httpd_port = 62354
tasks = WakeOnLan,Deploy,Collect,Inventory
```
I can see in GLPI on the Computer Form that the agent did send the inventory with no problem
```
UserAgent: GLPI-Agent_v1.5-git6b8b966f
Dernier inventaire: 2023-05-03 14:21
```

**Reply by g-bougard** [reply]
Importing manually the partial json from your output works on a 10.0.7 test instance:
![image](https://user-images.githubusercontent.com/12514489/235937623-e6fa9dfb-ba3e-4930-8118-6ea6802ed1b5.png)
So it must be processed by the server as expected.

Anyway I'll add the module so the next nightly build should include it.

**Reply by tiangao88** [reply]
Hi @g-bougard 

First of all thank you for coding this so fast.

I did another test to confirm that the agent communicates with the server and that there is no rule problem that prevents importing computer data
<img width="1271" alt="image" src="https://user-images.githubusercontent.com/53409436/235967238-b32c4038-9c9c-479e-be70-8973f658b54c.png">

Unfortunately there is no Antivirus information imported at all on the Computer Form
<img width="1172" alt="image" src="https://user-images.githubusercontent.com/53409436/235968299-53fcf8af-04bb-4067-9094-5974f8fb25ea.png">

I am wondering whether the antivirus information was really sent to the server.
So I modified the configuration of my GLPI-Agent as follow
```
server = https://glpi.myserver.com/fonrt/inventory.php
local = /Users/macuser/Downloads/DUMP-GLPI-AGENT.txt
```
I noticed it actually does 2 inventories one after the other but I guess the content is the same.

Anyway in the DUMP file there is no Antivirus section, the keyword does not appear in the file.
I have made an extract of all sections mentioning the keyword Defender that you will find attached.
[DUMP-EXTRACT.txt](https://github.com/glpi-project/glpi-agent/files/11384721/DUMP-EXTRACT.txt)

Is there something wrong in my agent configuration file?

**Reply by g-bougard** [reply]
Maybe `mdatp` command is not in the PATH when run from the service.
Can you share the output of `which mdatp` ?

**Reply by tiangao88** [reply]
So I did another test with 2 configuration files for the agent. One that sends data to the server and a separate one that sends data to a file.

Here is the config file for sending data to a file
```
local = /Users/macuser/Downloads/GLPI-ANTIVIRUS-INFO.txt
no-category =
accesslog,battery,bios,controller,cpu,database,drive,environment,firewall,hardware,input,licenseinfo,local_group,local_user,lvm,memory,modem,monitor,network,os,port,printer,process,provider,psu,registry,remote_mgmt,ruder,slot,software,sound,storage,usb,user,video,virtualmachine

no-ssl-check = 0
timeout = 30
delaytime = 3600
httpd_port = 62354
tasks = Inventory
```

By excluding all categories except antivirus I was expecting to get the same data as `glpi-inventory --partial antivirus` in the file GLPI-ANTIVIRUS-INFO.txt
But the content is the same as my previous test, and still no antivirus section.

**Reply by g-bougard** [reply]
> Maybe `mdatp` command is not in the PATH when run from the service.
> Can you share the output of `which mdatp` ?

Did you miss that comment ?

**Reply by tiangao88** [reply]
Here it is
```
which mdatp
/usr/local/bin/mdatp
```

**Reply by tiangao88** [reply]
Hi @g-bougard 

So I modified AntiVirus.pm like this `command => '/usr/local/bin/mdatp health --output json'`

The agent still communicates well with the server
<img width="993" alt="image" src="https://user-images.githubusercontent.com/53409436/236132064-2791b814-e8e3-4260-99af-f5e707d29b93.png">

But there is still no anvitivirus information transfered to the server
<img width="626" alt="image" src="https://user-images.githubusercontent.com/53409436/236132356-ae06414b-d680-4ef3-808b-ab9690853fe1.png">

**Reply by g-bougard** [reply]
Can you confirm where you put the `AntiVirus.pm` perl module ?
Anyway, you should probably enable `debug = 2` in configuration, restart the service, force an inventory and check the log `/var/log/glpi-agent.log` to see if the module has been enabled and if it has run it.
In GLPI, you can download the json and check if it contains the `antivirus` entry.

**Reply by g-bougard** [reply]
Hi @tiangao88 
indeed, you probably forgot to also fix L.17 with:
```perl
    return canRun('/usr/local/bin/mdatp');
```

**Reply by tiangao88** [reply]
So I fixed line 17 the file `/Applications/GLPI-Agent/agent/GLPI/Agent/Task/Inventory/MacOS/AntiVirus.pm` and added `debug = 2` in the agent configuration

In `/var/log/glpi-agent.log` I found the following

```
"antivirus": [
         {
            "base_version": "1.389.207.0",
            "company": "Microsoft",
            "enabled": true,
            "expiration": "2023-09-06",
            "name": "Microsoft Defender",
            "uptodate": true,
            "version": "101.98.30"
         }
      ],
```

And in GLPI tadaaaa !
<img width="1566" alt="image" src="https://user-images.githubusercontent.com/53409436/236311241-0249fa82-0017-4bd8-bc48-08a1db7343fd.png">

So the problem was indeed that full path `/usr/local/bin/mdatp` should be used everywhere in `AntiVirus.pm`

I will test the nightly version after you have confirmed this module is included.

**Reply by g-bougard** [reply]
Great !
I confirm you the latest nightly includes the same fix.
So you still can try it and confirm this works as expected.

**Reply by tiangao88** [reply]
hi @g-bougard 

I confirm that with the nightly build of May 5th I can see antivirus information updating in GLPI Server.
<img width="973" alt="image" src="https://user-images.githubusercontent.com/53409436/236443769-5a07264a-ed9b-489a-ab9b-14f3acab1c4c.png">

We will wait for GLPI-Agent 1.5 to make a full deploy on all our Macs.
Thank you for adding this module!

**Comment by mrb-x** [comment]
Hello @g-bougard,
I wanted inventoried SentinelOne EPP on macOS.
I found this discussion, so I created ([SentinelOne.pm.zip](https://github.com/user-attachments/files/16393084/SentinelOne.pm.zip)) (a mimic version of Cortex.pm from GLPI-Agent 1.10).
For now the agent report the Version and a Value of signature in base_version and state of Protection.
I'm not sure about the reliability of the datas sources for now :-)

Thanks

**Comment by g-bougard** [comment]
Hello @mrb-x 

good job, can I ask you to better make a PR so you'll be the referenced as the code author ? I only see one point which is not optimized: you're calling 2 times the same command with "status" option where you can get the output of one launch and analyze it 2 times. You can try to change lines 55 to 77 with this shorter code:
```perl
    # Support file case for unittests if basefile is provided
    if (empty($params{basefile})) {
        $params{command} = "\"$command\" status";
    } else {
        $params{file} = $params{basefile}."-status";
    }
    my @status_lines = getAllLines(%params);
    my ($staticSignatures) = grep { /^\s.*staticSignatures:/i } @status_lines;
    $antivirus->{BASE_VERSION} = $1
        if $staticSignatures && $staticSignatures =~ /^\s.*staticSignatures:\s.*(\([0-9]+\))/i
    $antivirus->{ENABLED} = 1 if grep { /^\s.*Protection.*enabled$/i } @status_lines;
```

In the PR comment, I would also like to see the outputs of the command for the 2 cases with option `version` & `status` so I can add tests after the merge. If you don't know how to or don't want to make a PR. Can you provide these outputs here ?

In the outputs, did you see any relevant useful datas ? The format is described here: https://github.com/glpi-project/inventory_format/blob/2b46a1ee491fefcec668acede800cef5fb276618/inventory.schema.json#L60
The other relevant fields could be: `UPTODATE`, `BASE_CREATION` & `EXPIRATION`
Anyway, can you eventually provide a link to an official and public SentinelOne EPP documentation ?
